File: sdk/typescript/src/credentials.ts
import { KYCVaultClient, VerifiableCredential, VerifiablePresentation, Proof } from "./client";
import { CredentialError, ErrorCode } from "./errors";

export interface IssueCredentialParams {
  issuerDid: string;
  subjectDid: string;
  claims: Record<string, unknown>;
  expirationDate?: string;
  credentialType?: string[];
  proofPurpose?: string;
}

export interface VerifyCredentialParams {
  credential: VerifiableCredential;
  challenge?: string;
  domain?: string;
}

export interface CreatePresentationParams {
  credentials: VerifiableCredential[];
  holderDid: string;
  challenge?: string;
  domain?: string;
}

export class CredentialManager {
  private readonly client: KYCVaultClient;

  constructor(client: KYCVaultClient) {
    this.client = client;
  }

  async issue(params: IssueCredentialParams): Promise<VerifiableCredential> {
    this.validateSubject(params.subjectDid);

    const response = await this.client.issueCredential(
      params.issuerDid,
      params.subjectDid,
      params.claims,
      {
        expirationDate: params.expirationDate,
        credentialType: params.credentialType,
        proofPurpose: params.proofPurpose,
      },
    );

    return response.credential;
  }

  async verify(params: VerifyCredentialParams): Promise<boolean> {
    this.validateCredential(params.credential);

    if (params.credential.expirationDate) {
      const now = new Date();
      const exp = new Date(params.credential.expirationDate);
      if (now > exp) {
        throw CredentialError.expired(params.credential.id);
      }
    }

    const result = await this.client.verifyCredential(params.credential, {
      challenge: params.challenge,
      domain: params.domain,
    });

    return result.verified;
  }

  async verifyWithDetail(
    params: VerifyCredentialParams,
  ): Promise<{ verified: boolean; checks: { name: string; passed: boolean; message?: string }[] }> {
    this.validateCredential(params.credential);

    const result = await this.client.verifyCredential(params.credential, {
      challenge: params.challenge,
      domain: params.domain,
    });

    return {
      verified: result.verified,
      checks: result.checks,
    };
  }

  async createPresentation(
    params: CreatePresentationParams,
  ): Promise<VerifiablePresentation> {
    if (params.credentials.length === 0) {
      throw new CredentialError(
        "At least one credential is required to create a presentation",
        ErrorCode.VALIDATION_ERROR,
      );
    }

    return this.client.createPresentation(
      params.credentials,
      params.holderDid,
      {
        challenge: params.challenge,
        domain: params.domain,
      },
    );
  }

  async verifyPresentation(
    presentation: VerifiablePresentation,
    challenge?: string,
    domain?: string,
  ): Promise<boolean> {
    if (!presentation.verifiableCredential || presentation.verifiableCredential.length === 0) {
      throw new CredentialError(
        "Presentation must contain at least one credential",
        ErrorCode.VALIDATION_ERROR,
      );
    }

    const result = await this.client.verifyPresentation(presentation, {
      challenge,
      domain,
    });

    return result.verified;
  }

  async revoke(id: string, reason?: string): Promise<void> {
    const result = await this.client.revokeCredential(id, reason);
    if (!result.revoked) {
      throw new CredentialError(
        `Failed to revoke credential '${id}'`,
        ErrorCode.INTERNAL_ERROR,
        { credentialId: id },
      );
    }
  }

  async getStatus(id: string): Promise<{
    status: string;
    updatedAt: string;
    revokedAt?: string;
    revocationReason?: string;
  }> {
    return this.client.checkCredentialStatus(id);
  }

  async get(id: string): Promise<VerifiableCredential> {
    return this.client.getCredential(id);
  }

  async list(params?: {
    page?: number;
    pageSize?: number;
    subjectDid?: string;
    issuerDid?: string;
    status?: "valid" | "expired" | "revoked";
  }): Promise<VerifiableCredential[]> {
    const response = await this.client.listCredentials(params);
    return response.data;
  }

  async isExpired(credential: VerifiableCredential): Promise<boolean> {
    if (!credential.expirationDate) return false;

    // Check with the server first for authoritative answer
    try {
      const status = await this.getStatus(credential.id);
      return status.status === "expired";
    } catch {
      // Fallback to local check
      const now = new Date();
      const exp = new Date(credential.expirationDate);
      return now > exp;
    }
  }

  async isRevoked(credential: VerifiableCredential): Promise<boolean> {
    try {
      const status = await this.getStatus(credential.id);
      return status.status === "revoked";
    } catch {
      return false;
    }
  }

  async selectByType(
    credentials: VerifiableCredential[],
    type: string,
  ): Promise<VerifiableCredential[]> {
    return credentials.filter((c) => c.type.includes(type));
  }

  async selectNonExpired(
    credentials: VerifiableCredential[],
  ): Promise<VerifiableCredential[]> {
    const now = new Date();
    return credentials.filter((c) => {
      if (!c.expirationDate) return true;
      return now <= new Date(c.expirationDate);
    });
  }

  private validateSubject(did: string): void {
    if (!did || !did.startsWith("did:")) {
      throw new CredentialError(
        `Invalid subject DID: '${did}'`,
        ErrorCode.VALIDATION_ERROR,
        { subjectDid: did },
      );
    }
  }

  private validateCredential(credential: VerifiableCredential): void {
    if (!credential.id) {
      throw new CredentialError("Credential must have an id", ErrorCode.VALIDATION_ERROR);
    }
    if (!credential.issuer) {
      throw new CredentialError("Credential must have an issuer", ErrorCode.VALIDATION_ERROR);
    }
    if (!credential.credentialSubject?.id) {
      throw new CredentialError(
        "Credential subject must have an id",
        ErrorCode.VALIDATION_ERROR,
      );
    }
  }
}

export function createCredential(
  issuer: string,
  subjectId: string,
  claims: Record<string, unknown>,
  options?: {
    id?: string;
    types?: string[];
    expirationDate?: string;
  },
): VerifiableCredential {
  const credential: VerifiableCredential = {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://www.w3.org/2018/credentials/examples/v1",
    ],
    id: options?.id || `urn:uuid:${crypto.randomUUID()}`,
    type: options?.types || ["VerifiableCredential"],
    issuer,
    issuanceDate: new Date().toISOString(),
    expirationDate: options?.expirationDate,
    credentialSubject: {
      id: subjectId,
      ...claims,
    },
  };

  return credential;
}

export function addProof(
  credential: VerifiableCredential,
  proof: Proof,
): VerifiableCredential {
  return {
    ...credential,
    proof,
  };
}

export function toJSON(credential: VerifiableCredential): string {
  return JSON.stringify(credential, null, 2);
}

export function fromJSON(json: string): VerifiableCredential {
  return JSON.parse(json) as VerifiableCredential;
}

export function isVerifiableCredential(obj: unknown): obj is VerifiableCredential {
  if (typeof obj !== "object" || obj === null) return false;
  const vc = obj as Record<string, unknown>;
  return (
    typeof vc.id === "string" &&
    typeof vc.issuer === "string" &&
    Array.isArray(vc.type) &&
    typeof vc.credentialSubject === "object" &&
    vc.credentialSubject !== null &&
    typeof (vc.credentialSubject as Record<string, unknown>).id === "string"
  );
}

File: sdk/flutter/lib/credentials.dart
import 'kyc_vault.dart';

class CredentialManager {
  final KYCVaultClient _client;

  CredentialManager(this._client);

  void _validateSubject(String did) {
    if (!did.startsWith('did:')) {
      throw CredentialError("Invalid subject DID: '$did'", 'VALIDATION_ERROR');
    }
  }

  void _validateCredential(Map<String, dynamic> credential) {
    if (credential['id'] == null) {
      throw CredentialError('Credential must have an id', 'VALIDATION_ERROR');
    }
    if (credential['issuer'] == null) {
      throw CredentialError(
          'Credential must have an issuer', 'VALIDATION_ERROR');
    }
    if (credential['credentialSubject'] == null ||
        credential['credentialSubject']['id'] == null) {
      throw CredentialError(
          'Credential subject must have an id', 'VALIDATION_ERROR');
    }
  }

  Future<Map<String, dynamic>> issue({
    required String issuerDid,
    required String subjectDid,
    required Map<String, dynamic> claims,
    String? expirationDate,
    List<String>? credentialType,
    String? proofPurpose,
  }) async {
    _validateSubject(subjectDid);
    final response = await _client.issueCredential(
      issuerDid,
      subjectDid,
      claims,
      expirationDate: expirationDate,
      credentialType: credentialType,
      proofPurpose: proofPurpose,
    );
    return response['credential'] as Map<String, dynamic>;
  }

  Future<bool> verify(
    Map<String, dynamic> credential, {
    String? challenge,
    String? domain,
  }) async {
    _validateCredential(credential);

    if (credential['expirationDate'] != null) {
      final now = DateTime.now();
      final exp = DateTime.parse(credential['expirationDate'] as String);
      if (now.isAfter(exp)) {
        throw CredentialError(
            "Credential '${credential['id']}' is expired", 'EXPIRED_CREDENTIAL');
      }
    }

    final result = await _client.verifyCredential(
      credential,
      challenge: challenge,
      domain: domain,
    );
    return result['verified'] as bool;
  }

  Future<Map<String, dynamic>> verifyWithDetail(
    Map<String, dynamic> credential, {
    String? challenge,
    String? domain,
  }) async {
    _validateCredential(credential);
    final result = await _client.verifyCredential(
      credential,
      challenge: challenge,
      domain: domain,
    );
    return {
      'verified': result['verified'],
      'checks': result['checks'],
    };
  }

  Future<Map<String, dynamic>> createPresentation({
    required List<Map<String, dynamic>> credentials,
    required String holderDid,
    String? challenge,
    String? domain,
  }) async {
    if (credentials.isEmpty) {
      throw CredentialError(
        'At least one credential is required to create a presentation',
        'VALIDATION_ERROR',
      );
    }
    return _client.createPresentation(
      credentials,
      holderDid,
      challenge: challenge,
      domain: domain,
    );
  }

  Future<bool> verifyPresentation(
    Map<String, dynamic> presentation, {
    String? challenge,
    String? domain,
  }) async {
    final vc = presentation['verifiableCredential'];
    if (vc == null || (vc as List).isEmpty) {
      throw CredentialError(
        'Presentation must contain at least one credential',
        'VALIDATION_ERROR',
      );
    }
    final result = await _client.verifyPresentation(
      presentation,
      challenge: challenge,
      domain: domain,
    );
    return result['verified'] as bool;
  }

  Future<void> revoke(String id, {String? reason}) async {
    final result = await _client.revokeCredential(id, reason: reason);
    if (result['revoked'] != true) {
      throw CredentialError("Failed to revoke credential '$id'", 'INTERNAL_ERROR');
    }
  }

  Future<Map<String, dynamic>> getStatus(String id) async {
    return _client.checkCredentialStatus(id);
  }

  Future<Map<String, dynamic>> get(String id) async {
    return _client.getCredential(id);
  }

  Future<List<Map<String, dynamic>>> list({
    int? page,
    int? pageSize,
    String? subjectDid,
    String? issuerDid,
    String? status,
  }) async {
    final response = await _client.listCredentials(
      page: page,
      pageSize: pageSize,
      subjectDid: subjectDid,
      issuerDid: issuerDid,
      status: status,
    );
    return List<Map<String, dynamic>>.from(response['data'] ?? []);
  }

  Future<bool> isExpired(Map<String, dynamic> credential) async {
    if (credential['expirationDate'] == null) return false;
    try {
      final status = await getStatus(credential['id'] as String);
      return status['status'] == 'expired';
    } catch (_) {
      final now = DateTime.now();
      final exp = DateTime.parse(credential['expirationDate'] as String);
      return now.isAfter(exp);
    }
  }

  Future<bool> isRevoked(Map<String, dynamic> credential) async {
    try {
      final status = await getStatus(credential['id'] as String);
      return status['status'] == 'revoked';
    } catch (_) {
      return false;
    }
  }

  Future<List<Map<String, dynamic>>> selectByType(
    List<Map<String, dynamic>> credentials,
    String type,
  ) async {
    return credentials.where((c) {
      final types = List<String>.from(c['type'] ?? []);
      return types.contains(type);
    }).toList();
  }

  Future<List<Map<String, dynamic>>> selectNonExpired(
    List<Map<String, dynamic>> credentials,
  ) async {
    final now = DateTime.now();
    return credentials.where((c) {
      if (c['expirationDate'] == null) return true;
      final exp = DateTime.parse(c['expirationDate'] as String);
      return !now.isAfter(exp);
    }).toList();
  }
}

class CredentialError implements Exception {
  final String message;
  final String code;
  final Map<String, dynamic>? details;

  CredentialError(this.message, this.code, [this.details]);

  @override
  String toString() => 'CredentialError($code): $message';
}

Map<String, dynamic> createCredential(
  String issuer,
  String subjectId,
  Map<String, dynamic> claims, {
  String? id,
  List<String>? types,
  String? expirationDate,
}) {
  final now = DateTime.now().toIso8601String();
  return {
    '@context': [
      'https://www.w3.org/2018/credentials/v1',
      'https://www.w3.org/2018/credentials/examples/v1',
    ],
    'id': id ?? 'urn:uuid:${_uuidV4()}',
    'type': types ?? ['VerifiableCredential'],
    'issuer': issuer,
    'issuanceDate': now,
    if (expirationDate != null) 'expirationDate': expirationDate,
    'credentialSubject': {
      'id': subjectId,
      ...claims,
    },
  };
}

Map<String, dynamic> addProof(
  Map<String, dynamic> credential,
  Map<String, dynamic> proof,
) {
  return {
    ...credential,
    'proof': proof,
  };
}

String credentialToJSON(Map<String, dynamic> credential) {
  return credential.toString();
}

Map<String, dynamic> credentialFromJSON(String json) {
  return json as Map<String, dynamic>;
}

bool isVerifiableCredential(dynamic obj) {
  if (obj is! Map<String, dynamic>) return false;
  return obj['id'] is String &&
      obj['issuer'] is String &&
      obj['type'] is List &&
      obj['credentialSubject'] is Map &&
      obj['credentialSubject']['id'] is String;
}

String _uuidV4() {
  final now = DateTime.now().millisecondsSinceEpoch;
  return '${now}-${now.hashCode}';
}
